### PR TITLE
[Relay] Fix invalid shape function for "copy" operator

### PR DESCRIPTION
**Pull request by mbs-octoml**
The `@script` shape function for `copy` was ill-formed, resulting in a TIR shape
function which did not assign to its output. That in turn caused either OOM or
assert fails as uninitialized shape dimensions worked their way downstream.
That fix is in `python/tvm/relay/op/tensor.py`.

Special thanks to @electriclilies  for helping me with the scalar vs tensor switch
in the shape function. It turns out `copy` is used for scalars in some models.

Everything else is for testing and debugging as I tracked
this down.

[This is CORE-112 in OctoML JIRA.]

### CODE REVIEW

File: include/tvm/runtime/debug.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ * shapes and dtypes, but also their contents if 'small' and on the \p host_device (mostly so that
+ * we can see dynamic shapes as they are computed). For \p adts we show the ADT fields. For
+ * \p objects we dispatch to one of the above as appropriate.
+ */
+void AppendNDArray(std::ostream& os, const NDArray& nd_array, const DLDevice& host_device,
+                   bool show_content = true);
+void AppendADT(std::ostream& os, const ADT& adt, const DLDevice& host_device,
+               bool show_content = true);
+void AppendRuntimeObject(std::ostream& os, const ObjectRef& object, const DLDevice& host_device,
+                         bool show_content = true);
```
Comment (electriclilies) [inline include/tvm/runtime/debug.h:47]: Nice, this will make debugging shape functions easier!

File: src/relay/backend/te_compiler_cache.cc [hunk 2]
```diff
@@ -635,20 +640,16 @@ class MakeShapeFunc : public backend::MemoizedExprTranslator<Array<te::Tensor>>
     // Get output ndims
     auto ret_type = call_node->checked_type();
     Array<IndexExpr> out_ndims;
-    if (const auto* ttype = ret_type.as<TensorTypeNode>()) {
+    for (const auto& ttype : FlattenTupleType(ret_type)) {
       out_ndims.push_back(IntImm(DataType::Int(32), ttype->shape.size()));
```
Comment (electriclilies) [inline src/relay/backend/te_compiler_cache.cc:644]: Is this change related to fixing the copy operator or just a general improvement?

File: src/target/compilation_config.cc [hunk 3]
```diff
@@ -61,47 +61,47 @@ void CompilationConfigNode::EstablishDefaultSEScopes(const transform::PassContex
   if (host_target.defined()) {
     CHECK(!host_target->host.defined()) << "Host targets are not expected to have hosts";
     host_device_type = static_cast<DLDeviceType>(host_target->kind->device_type);
-    DLOG(INFO) << "Using the given host target " << host_target->ToDebugString()
-               << " of device type " << host_device_type << " for the host target";
+    VLOG(1) << "Using the given host target " << host_target->ToDebugString() << " of device type "
+            << host_device_type << " for the host target";
```
Comment (electriclilies) [inline src/target/compilation_config.cc:65]: Have we discussed VLOG vs DLOG? It seems like people are still adding DLOGs. Might be worth discussing unifying around one debugging system at some point in the future.

File: tests/python/relay/dyn/test_dynamic_op_level3.py [hunk 4]
```diff
@@ -410,5 +418,75 @@ def verify_sparse_fill_empty_rows(
...
+        }
+        """
+    )
+    x_data = 3
+    y_data = np.random.rand(7).astype("int32")
+    expected = np.concatenate((y_data, np.expand_dims(x_data, axis=0)))
+    check_on_vm(target, dev, [x_data, y_data], expected, mod)
+
+
+def test_dyn_expand_dims():
```
Comment (electriclilies) [inline tests/python/relay/dyn/test_dynamic_op_level3.py:457]: There's another test_dyn_expand_dims in this file, possibly bad rebase?

**Review by electriclilies — APPROVED** [review]
Overall LGTM! I have a few nitty questions though

File: include/tvm/runtime/debug.h (hunk 1 above)
Comment (mbs-octoml) [reply include/tvm/runtime/debug.h:47]: Yeah, eventually I'd like to have spans in the VM instruction stream too.

File: src/relay/backend/te_compiler_cache.cc (hunk 2 above)
Comment (mbs-octoml) [reply src/relay/backend/te_compiler_cache.cc:644]: Sorry, unrelated. Cleaning up flattening as I go.

File: src/target/compilation_config.cc (hunk 3 above)
Comment (mbs-octoml) [reply src/target/compilation_config.cc:65]: Yeah. In this case these were my own DLOGs and they were too verbose even for me.

In general getting folks to think about debugability when writing code in the first place would be a fine thing. Eg I don't know if there's any way to see TE, or Schedules.

File: tests/python/relay/dyn/test_dynamic_op_level3.py (hunk 4 above)
Comment (mbs-octoml) [reply tests/python/relay/dyn/test_dynamic_op_level3.py:457]: Oh, no just ignorance on my part -- no need to test it again! thx.

**Comment by mbs-octoml** [comment]
Thanks @electriclilies . Here goes another 6 hours of CI!

**Comment by mbs-octoml** [comment]
Dang, stepped on my own toes in #9759. Here we go again.

**Review by masahi — APPROVED** [review]

**Pull request merged**